lib/charms/traefik_k8s/v0/ingress_per_unit.py

```diff
@@ -419,10 +437,60 @@ def is_failed(self, relation: Relation = None) -> bool:
 
         return False
 
-    def get_request(self, relation: Relation):
-        """Get the IngressRequest for the given Relation."""
-        provider_app_data, requirer_unit_data = self._fetch_relation_data(relation)
-        return IngressRequest(self, relation, provider_app_data, requirer_unit_data)
+    def is_unit_ready(self, relation: Relation, unit: Unit) -> bool:
+        """Report whether the given unit has shared data in its unit data bag."""
+        # sanity check: this should not occur in production, but it may happen
+        # during testing: cfr https://github.com/canonical/traefik-k8s-operator/issues/39
+        assert (
+            unit in relation.units
+        ), "attempting to get ready state for unit that does not belong to relation"
+        if relation.data.get(unit, {}).get("data"):
+            # TODO consider doing schema-based validation here
+            return True
+        return False
+
+    def get_data(self, relation: Relation, unit: Unit, validate: bool = False) -> "RequirerData":
```

Comment:
Perhaps for another PR, but should the validation be optional? I'm not sure what that gains us on the provider side -- I thought we spoke about wrapping the `jsonschema` import in a try/except to potentially allow charms that don't add that dependency to function, but with a warning. Perhaps here that could reflected in an automatic choice as to whether validation should occur?

Reply:
Good point. I uniformed the behaviour and put a global DO_VALIDATION flag depending on the presence of `jsonschema`. Should make more sense now; also it fires a nice warning if it is not installed, and will work with and without.